Add command line support for creating, building and running containerized integrations locally

doru1004 opened the pull request:
<!-- Description -->

This patch adds command line options for creating, building and running local images.

A new local command, `local create` has been added to construct images.

The first image kind it can produce is a base image to be used as a starter for all integration images. Example of building a base image, the image will be saved under the default name of `integration-base-image`:

```
kamel local create --base-image --container-registry <registry-name>
```

The image can then be used as a base for creating and building integration images, the image will be used automatically every time an integration image is created. If a base image isn't available it will be created automatically. Example of building an integration image (without running it):

```
kamel local create --image <registry-name>/<my-image-name> -d <dep> -p <prop> --property-file <file> route1.java route2.yaml
```

The resulting image can then be run locally using the following invocation:

```
kamel local run --image <registry-name>/<my-image-name>
```

Local run has been enhanced with containerization capabilities i.e. it can now create, build and run container images locally. The following command containerizes and runs the integration (i.e. performs all previous steps automatically: base image creation and building, integration image creation and building, integration image running):

```
kamel local run --containerize --image <registry-name>/<my-image-name> -d <dep> -p <prop> --property-file <file> route1.java route2.yaml
```


<!--
Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". 

You can (optionally) mark this PR with labels "kind/bug" or "kind/feature" to make sure
the text is added to the right section of the release notes. 
-->

**Release Note**
```release-note
Added ability to package and run integrations as container images
```

nicolaferraro commented on pkg/cmd/local_create.go:
```diff
@@ -0,0 +1,168 @@
+
+	cmd.Flags().Bool("base-image", false, "Create base image used as a starting point for any integration.")
+	cmd.Flags().String("image-name", "", "Integration image name.")
+	cmd.Flags().String("docker-registry", "", "Docker registry to store intermediate images.")
```
I'd prefer using the word "container" instead of Docker here.

Also, I wonder if it would be better to have a single parameter for the image name and registry. Simplifying the command as: `kamel local create it.yaml --image docker.io/myrepo/myimage`, i.e. treat is as a tag.

doru1004 replied:
The reason why I have separated the two flags is because the base image doesn't need a name (name is always the same), it only needs a registry. So I'd have to keep the flag around for that reason and rename it to `container-registry` and then for integration images I can just support the `--image` flag like you suggested and make it illegal to use the registry flag when an integration is being built. What do you think?

nicolaferraro replied:
Sounds good

doru1004 replied:
Cool, I implemented that strategy.

nicolaferraro commented on pkg/cmd/local_run.go:
```diff
@@ -55,31 +55,60 @@ func newCmdLocalRun(rootCmdOptions *RootCmdOptions) (*cobra.Command, *localRunCm
 
+	cmd.Flags().Bool("containerize", false, "Run integration in a local container.")
+	cmd.Flags().String("image-name", "", "Integration image name.")
+	cmd.Flags().String("docker-registry", "", "Docker registry to store intermediate images.")
```
Same comment as before.. in all places

doru1004 replied:
Done.

nicolaferraro commented on pkg/util/docker/docker.go:
```diff
@@ -0,0 +1,141 @@
+	dockerFile := []string{}
+
+	// Base image is a java-only image since the integration command is just a java command.
+	dockerFile = append(dockerFile, FROM("adoptopenjdk/openjdk11:alpine"))
```
You should use the base image from `defaults.go`

doru1004 replied:
Done

nicolaferraro commented on pkg/util/util.go:
```diff
@@ -35,6 +36,23 @@ import (
+const DefaultRoutesDirectoryName = "routes"
+
+// DefaultWorkingDirectoryName --
+const DefaultWorkingDirectoryName = "workspace"
```
Can we use the same structure found in the kubernetes image? There's a small structure under `/deployments` there, plus many of other things are present in `/etc/camel`.

The reason why this is important is that many users would like to use this command to freeze a container image that is then used in test and production, so it would be nice if a base image built here can be used as base image in kube as well

doru1004 replied:
Done

nicolaferraro commented on pkg/cmd/local_create.go:
```diff
@@ -0,0 +1,168 @@
+		},
+	}
+
+	cmd.Flags().Bool("base-image", false, "Create base image used as a starting point for any integration.")
```
I like the idea of building a base-image for an integration (without embedding the sources) or an image with source embedded.

However, I've tested this with an integration (base-image=false) and sources are not included. If this is the idea, I think there's something still not working correctly.

nicolaferraro replied:
I re-read the description and I think I didn't get it correctly...

doru1004 replied:
If there are any further questions about this please let me know :)

nicolaferraro reviewed (changes requested):
Thanks @doru1004 , it looks really promising. Just some comments to improve it.

nicolaferraro commented:
Another issue is that `kamel local create --base-image` creates a maven structure in the current directory (pom.xml, src and target dirs).

doru1004 commented:
> Another issue is that `kamel local create --base-image` creates a maven structure in the current directory (pom.xml, src and target dirs).

I fixed this. There were several other files and folders emitted and now no intermediate files/folders are emitted in the current directory.

nicolaferraro commented:
Seems there is some other issue, I can't get `kamel local run examples/routes.groovy` to work (it does not print the periodic log message).

doru1004 commented:
> Seems there is some other issue, I can't get `kamel local run examples/routes.groovy` to work (it does not print the periodic log message).

It looks like this has nothing to do with the latest changes, it probably has to do with redirecting the output of the command.

doru1004 commented:
I have updated the patch to fix the reported problems. I have also changed the directories used inside the image to be the same ones used by Kubernetes! :)

doru1004 commented:
@nicolaferraro not sure why the knative test failed, it looks like there is a similar failure in master.

doru1004 commented:
Rebased on latest master.

nicolaferraro commented on pkg/cmd/util_containerization.go:
```diff
@@ -0,0 +1,178 @@
+	fmt.Printf("Executing: " + strings.Join(cmd.Args, " ") + "\n")
+
+	// Run the command.
+	if err := cmd.Run(); err != nil {
```
Should you set the stdout and stderr here? It works but I see no logs

doru1004 replied:
Can definitely do that. Are there some guidelines as to how std out/err should be managed?

doru1004 replied:
I've added some simple redirection of stdout and stderr.

doru1004 commented:
I have also added the capability to include environment variable in modeline options:

test.yaml
```
# camel-k: property=camel.component.slack.webhookUrl=${SLACK_WEBHOOK}
```

When the modeline option is extracted, it will be added to the command line like this:

```
kamel local run --property=camel.component.slack.webhookUrl=hooks.slack.com/AAA/BBB/CCC
```

If the variable is not exported or has no value an error will be thrown.

doru1004 commented:
Rebased on latest master.
